vllm/entrypoints/openai/serving.py

```diff
@@ -0,0 +1,935 @@
+    CompletionRequest, CompletionResponse, CompletionResponseChoice,
+    CompletionResponseStreamChoice, CompletionStreamResponse,
+    ChatCompletionRequest, ChatCompletionResponse,
+    ChatCompletionResponseChoice, ChatCompletionResponseStreamChoice,
+    ChatCompletionStreamResponse, ChatMessage, DeltaMessage, LogProbs,
+    ModelCard, ModelList, ModelPermission, UsageInfo, ChatCompletionToolParam,
+    ToolCallsDelta, ToolCallsMessage, FunctionCall,
+    ChatCompletionAssistantMessage, ChatCompletionToolMessage, ErrorResponse)
+from vllm.outputs import RequestOutput
+from vllm.sampling_params import SamplingParams
+from vllm.transformers_utils.tokenizer import get_tokenizer
+from vllm.utils import random_uuid
+
+logger = init_logger(__name__)
+
+
+class OpenAIToolsPrompter:
+    """
+    https://platform.openai.com/docs/assistants/tools
+    """
+
+    def __init__(self):
+        pass
+
+    @classmethod
+    def func_call_token_pre(cls) -> str:
+        return "!"
+
+    @classmethod
+    def func_call_token_size(cls) -> int:
+        return 15
+
+    @classmethod
+    def func_call_token(cls) -> str:
+        return "!function_call:"
+
+    def content_from_assistant(self,
+                               message: ChatCompletionAssistantMessage) -> str:
+        text = ""
+        for call in message.tool_calls:
+            text += call.id + " was called with arguments : " + str(
+                call.function.arguments) + "\n"
+        if message.content is None:
+            return text
+        else:
+            return message.content + "\n" + text
+
+    def content_from_tool(self, message: ChatCompletionToolMessage) -> str:
+        return message.tool_call_id + " -> " + message.content
+
+    def inject_prompt(self, request: ChatCompletionRequest):
+        """ Tested with :
+                https://huggingface.co/mlabonne/NeuralHermes-2.5-Mistral-7B/discussions/3 """
+        if request.tool_choice is not None and request.tools is not None and request.tool_choice == "auto":
+            tools_list: [ChatCompletionToolParam] = request.tools
+            if len(tools_list):
+                text_inject = (
+                    "Your task is to call a function when needed. "
+                    "You will be provided with a list of functions. "
+                    "\n\nAvailable functions:\n")
+                for tool in tools_list:
+                    if tool.type == "function":
+                        text_inject += "\n" + tool.function.name
+                        if tool.function.description is not None:
+                            text_inject += " - " + tool.function.description
+                        if tool.function.parameters is not None:
+                            schema = json.dumps(tool.function.parameters,
+                                                indent=4)
+                            text_inject += f"```\njsonschema\n{schema}\n```"
+                text_inject += (
```

Comment:
Should this be hard coded? It seems like depending on how the model was trained for function calling, the specific injected text could have a large impact on performance of the model. Seems like a good idea to allow this to be configurable IMO.

Reply:
Following what was suggested by @Tostino , I think that it could be very handly to incorporate a `tool_template.json` path to personalize the tool response (probably preserving the current as default)

Particularly, the way the json_schema_params in constructed IMO it's ok. I wouldn't change it. 
Nevertheless, incorporating a `tool_template.json` like:

```json
{
    "prefix": "str",
    "suffix": "str",
    "order": "top"
    "func_call_token": "str"
}
```
could provide enough freedom in regards of models' adaptation when `inject_prompt` is performed. ( For ease of understanding please refer to image below )

What's do you think about @simon-mo ?

Image:
![tool_template](https://github.com/vllm-project/vllm/assets/37890346/997f2fa6-a727-47a2-bc5a-6e8f6d69a50d)

Reply:
It's a good idea and easy to implement.

If we need to implement templates, it means that ideally we should remove all serving.py instructions.
So, we need to define a path to store a default tool_template.json.

The examples/ folder is not suitable.

There are four options:
- Hardcode it as a string in serving.py (which I don't like)
- The vllm/entrypoints/openai/ folder (I'm undecided)
- Define a vllm/data/ folder or something similar (I'm undecided)
- Add a command-line argument to allow users to define a configuration directory (or $HOME/.config/vllm/openai by default) and store data there.

I prefer the last option, knowing that a general-purpose directory might be used for another type of templates.
Besides function calls, LLMs (starting with ChatGPT) can be extended by embeddings (which could potentially be implemented in the future in vLLM), dynamic Python interpreter, etc...
Similarly, I am not sure if this is the primary goal of vLLM, but having a compatible OpenAI server with most features ready to use seems as important as inference performance!

The OpenAI's documentation gives an idea of what could potentially be implemented.
https://platform.openai.com/docs/assistants/tools

Reply:
I'd follow the same approach as `chat-template`, command-line arg with a path.